File: scripts/collect-news.js
/**
 * Niger Info Veille - Script de Collecte d'Actualités (Collector)
 * Exécuté par Node.js (GitHub Actions ou localement)
 */

const fs = require('fs');
const path = require('path');
const crypto = require('crypto');
const https = require('https');
const http = require('http');
const { URL } = require('url');

// Charger Cheerio
let cheerio;
try {
  cheerio = require('cheerio');
} catch (e) {
  console.error("Erreur : La dépendance 'cheerio' n'est pas installée. Veuillez lancer 'npm install' dans le dossier scripts.");
  process.exit(1);
}

// Configuration
const USER_AGENT = "NigerInfoVeilleBot/1.0 (+contact@example.com)";
const MAX_ITEMS_PER_SOURCE = 10;
const MAX_TOTAL_ITEMS = 1000;
const MIN_TITLE_LENGTH = 30;
const TIMEOUT_MS = 10000; // 10 secondes
const IMAGE_ENRICH_LIMIT_PER_SOURCE = 6;

// Chemins des fichiers par rapport au script
const PATH_SOURCES = path.join(__dirname, '../data/sources.json');
const PATH_NEWS = path.join(__dirname, '../data/news.json');

// Utilitaire Sleep
const sleep = (ms) => new Promise((resolve) => setTimeout(resolve, ms));

/**
 * Télécharge le contenu HTML d'une page Web avec redirection et timeout.
 */
function fetchHtml(urlStr, depth = 0) {
  return new Promise((resolve, reject) => {
    if (depth > 3) {
      return reject(new Error('Trop de redirections'));
    }

    let urlObj;
    try {
      urlObj = new URL(urlStr);
    } catch (e) {
      return reject(new Error(`URL invalide: ${urlStr}`));
    }

    const client = urlObj.protocol === 'https:' ? https : http;
    const options = {
      hostname: urlObj.hostname,
      path: urlObj.pathname + urlObj.search,
      headers: {
        'User-Agent': USER_AGENT,
        'Accept': 'text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,*/*;q=0.8',
        'Accept-Language': 'fr,fr-FR;q=0.8,en-US;q=0.5,en;q=0.3',
      },
      timeout: TIMEOUT_MS,
      rejectUnauthorized: false // Ignore les certificats SSL invalides/expirés
    };

    const req = client.get(options, (res) => {
      // Gérer les redirections (301, 302, 307, 308)
      if (res.statusCode >= 300 && res.statusCode < 400 && res.headers.location) {
        const redirectUrl = new URL(res.headers.location, urlStr).toString();
        return resolve(fetchHtml(redirectUrl, depth + 1));
      }

      if (res.statusCode !== 200) {
        return reject(new Error(`Statut HTTP: ${res.statusCode}`));
      }

      let data = '';
      res.on('data', (chunk) => { data += chunk; });
      res.on('end', () => resolve(data));
    });

    req.on('error', (err) => reject(err));
    req.on('timeout', () => {
      req.destroy();
      reject(new Error('Timeout de requête atteint'));
    });
  });
}

function normalizeImageUrl(rawUrl, baseUrl) {
  if (!rawUrl || typeof rawUrl !== 'string') return '';

  const candidate = rawUrl
    .split(',')
    .pop()
    .trim()
    .split(/\s+/)[0];

  if (!candidate || candidate.startsWith('data:') || candidate.startsWith('blob:')) return '';

  try {
    const absoluteUrl = new URL(candidate, baseUrl).toString();
    const lower = absoluteUrl.toLowerCase();
    const rejectedTokens = [
      'logo', 'icon', 'favicon', 'avatar', 'sprite', 'pixel', 'tracking',
      'facebook', 'twitter', 'instagram', 'youtube', 'whatsapp'
    ];
    if (rejectedTokens.some(token => lower.includes(token)) || lower.endsWith('.svg')) return '';
    return absoluteUrl;
  } catch (e) {
    return '';
  }
}

function imageFromElement($, img, baseUrl) {
  if (!img || !img.length) return '';
  const attributes = [
    'data-src', 'data-original', 'data-lazy-src', 'data-image',
    'data-srcset', 'srcset', 'src'
  ];

  for (const attribute of attributes) {
    const imageUrl = normalizeImageUrl(img.attr(attribute), baseUrl);
    if (imageUrl) return imageUrl;
  }
  return '';
}

function extractPageImage(html, pageUrl) {
  try {
    const $ = cheerio.load(html);
    const metaSelectors = [
      'meta[property="og:image:secure_url"]',
      'meta[property="og:image"]',
      'meta[name="twitter:image"]',
      'meta[name="twitter:image:src"]',
      'link[rel="image_src"]'
    ];

    for (const selector of metaSelectors) {
      const element = $(selector).first();
      const imageUrl = normalizeImageUrl(element.attr('content') || element.attr('href'), pageUrl);
      if (imageUrl) return imageUrl;
    }

    const articleImage = $('article img, main img, .article img, .post img').first();
    return imageFromElement($, articleImage, pageUrl);
  } catch (e) {
    return '';
  }
}

/**
 * Génère un ID de news dédupliqué et prévisible.
 */
function generateNewsId(sourceName, title, url) {
  const hash = crypto.createHash('md5');
  hash.update(`${sourceName}-${title}-${url}`);
  return 'news-' + hash.digest('hex').substring(0, 12);
}

/**
 * Analyse le titre pour deviner la région concernée au Niger.
 */
function detectRegion(title) {
  const titleLower = title.toLowerCase();
  const regions = ['Niamey', 'Maradi', 'Zinder', 'Tahoua', 'Dosso', 'Tillabéri', 'Agadez', 'Diffa'];
  
  for (const region of regions) {
    if (titleLower.includes(region.toLowerCase())) {
      return region;
    }
  }
  return 'Niger';
}

/**
 * Estime la catégorie selon les mots-clés du titre.
 */
function estimateCategory(title, defaultCategory) {
  const t = title.toLowerCase();
  
  if (t.includes("appel d'offres") || t.includes("marché public") || t.includes("avis d'attribution") || t.includes("soumission")) {
    return "Marchés publics";
  }
  if (t.includes("santé") || t.includes("hôpital") || t.includes("vaccin") || t.includes("épidémie") || t.includes("maladie") || t.includes("polio")) {
    return "Santé";
  }
  if (t.includes("pluie") || t.includes("météo") || t.includes("climat") || t.includes("inondation") || t.includes("sécheresse") || t.includes("tempête")) {
    return "Météo";
  }
  if (t.includes("agriculture") || t.includes("semence") || t.includes("élevage") || t.includes("bétail") || t.includes("mil ") || t.includes("céréale")) {
    return "Agriculture";
  }
  if (t.includes("budget") || t.includes("finance") || t.includes("économie") || t.includes("banque") || t.includes("bceao") || t.includes("fmi") || t.includes("inflation")) {
    return "Économie";
  }
  if (t.includes("école") || t.includes("université") || t.includes("éducation") || t.includes("classe") || t.includes("scolaire")) {
    return "Éducation";
  }
  if (t.includes("humanitaire") || t.includes("déplacés") || t.includes("nutrition") || t.includes("réfugiés") || t.includes("aide alimentaire") || t.includes("ocha")) {
    return "Humanitaire";
  }
  if (t.includes("sécurité") || t.includes("militaire") || t.includes("attaque") || t.includes("défense") || t.includes("soldat") || t.includes("terrorisme")) {
    return "Sécurité";
  }
  if (t.includes("communiqué officiel") || t.includes("conseil des ministres") || t.includes("décret") || t.includes("gouvernement")) {
    return "Gouvernement";
  }
  if (t.includes("diaspora") || t.includes("ambassade") || t.includes("consulat")) {
    return "Diaspora";
  }
  
  return defaultCategory;
}

/**
 * Estime le score d'importance (1 à 5).
 */
function estimateImportance(title, sourceName, category) {
  const t = title.toLowerCase();
  
  // Critères de niveau 5 (Majeure)
  if (t.includes("urgent") || t.includes("alerte rouge") || t.includes("inondation majeure") || t.includes("attaque terroriste") || t.includes("coup d'état")) {
    return 5;
  }
  
  // Critères de niveau 4 (Très important)
  if (t.includes("communiqué") || t.includes("décret") || t.includes("alerte") || t.includes("sécurité") || t.includes("conseil des ministres")) {
    return 4;
  }
  
  // Critères de niveau 3 (Important)
  if (category === "Gouvernement" || category === "Marchés publics" || sourceName.includes("Présidence") || sourceName.includes("Gouvernement")) {
    return 3;
  }
  
  // Défaut
  return 2;
}

/**
 * Fonction principale du script de collecte.
 */
async function collect() {
  console.log('=== DÉBUT DE LA COLLECTE NIGER INFO VEILLE ===');
  const nowStr = new Date().toISOString();

  // 1. Lire les sources d'information
  if (!fs.existsSync(PATH_SOURCES)) {
    console.error(`Fichier des sources introuvable à : ${PATH_SOURCES}`);
    process.exit(1);
  }
  
  let sources = JSON.parse(fs.readFileSync(PATH_SOURCES, 'utf8'));
  console.log(`${sources.length} sources chargées.`);

  // 2. Lire les news actuelles pour fusion ultérieure
  let oldNews = [];
  if (fs.existsSync(PATH_NEWS)) {
    try {
      oldNews = JSON.parse(fs.readFileSync(PATH_NEWS, 'utf8'));
    } catch (e) {
      console.warn("Fichier news.json corrompu ou vide, réinitialisation.");
      oldNews = [];
    }
  }
  console.log(`${oldNews.length} articles actuellement en cache local.`);
  const oldNewsById = new Map(oldNews.map(item => [item.id, item]));

  const collectedNews = [];
  let successfulSourcesCount = 0;

  // Définition d'un worker individuel pour une source
  async function scrapeSource(source, index) {
    if (source.status !== 'active') {
      console.log(`[Source ${index}/${sources.length}] Ignorée : ${source.name} (inactive)`);
      return;
    }

    console.log(`[Source ${index}/${sources.length}] Début de l'analyse : ${source.name} (${source.url})`);
    
    try {
      const html = await fetchHtml(source.url);
      const $ = cheerio.load(html);
      const sourceItems = [];
      
      // Rechercher tous les liens dans la page
      $('a').each((idx, el) => {
        const href = $(el).attr('href');
        const text = $(el).text().replace(/\s+/g, ' ').trim();

        if (!href || !text) return;
        if (text.length < MIN_TITLE_LENGTH) return;

        const textLower = text.toLowerCase();
        const blacklist = [
          'accueil', 'contact', 'menu', 'facebook', 'twitter', 'linkedin', 'instagram', 'youtube',
          'lire plus', 'lire l\'article', 'read more', 'connexion', 's\'abonner', 'newsletter',
          'partager', 'télécharger', 'mentions légales', 'politique de confidentialité', 'qui sommes-nous'
        ];
        if (blacklist.some(b => textLower === b || textLower.startsWith(b))) return;

        if (href.startsWith('mailto:') || href.startsWith('tel:') || href.startsWith('javascript:') || href.startsWith('#')) {
          return;
        }

        let absoluteUrl;
        try {
          absoluteUrl = new URL(href, source.url).toString();
        } catch (e) {
          return;
        }

        // Tenter d'extraire l'image depuis le lien ou le conteneur de carte.
        let imageUrl = '';
        const imgEl = $(el).find('img').first();
        if (imgEl.length > 0) {
          imageUrl = imageFromElement($, imgEl, source.url);
        }

        if (!imageUrl) {
          const card = $(el).closest('article, li, .card, .post, .news-item, .item, .entry');
          if (card.length > 0) {
            imageUrl = imageFromElement($, card.find('img').first(), source.url);
          }
        }

        sourceItems.push({
          title: text,
          url: absoluteUrl,
          imageUrl: imageUrl
        });
      });

      // Dédupliquer
      const uniqueSourceItems = [];
      const seenUrls = new Set();
      for (const item of sourceItems) {
        if (!seenUrls.has(item.url)) {
          seenUrls.add(item.url);
          uniqueSourceItems.push(item);
        }
      }

      const selectedItems = uniqueSourceItems.slice(0, MAX_ITEMS_PER_SOURCE);

      // Les pages d'accueil utilisent souvent des images en arrière-plan ou ne
      // publient l'image que dans les métadonnées Open Graph de l'article.
      const itemsToEnrich = selectedItems
        .filter(item => !item.imageUrl)
        .slice(0, IMAGE_ENRICH_LIMIT_PER_SOURCE);

      await Promise.all(itemsToEnrich.map(async item => {
        try {
          const articleHtml = await fetchHtml(item.url);
          item.imageUrl = extractPageImage(articleHtml, item.url);
        } catch (e) {
          // Une image absente ne doit jamais faire échouer la collecte.
        }
      }));

      console.log(`[Source ${index}/${sources.length}] -> ${selectedItems.length} liens trouvés pour ${source.name}`);

      selectedItems.forEach(item => {
        const id = generateNewsId(source.name, item.title, item.url);
        const region = detectRegion(item.title);
        const category = estimateCategory(item.title, source.category);
        const importance = estimateImportance(item.title, source.name, category);
        const summary = `Cette information a été repérée depuis ${source.name}. Elle concerne : ${item.title}. Consultez la source originale pour lire le contenu complet.`;
        const previousItem = oldNewsById.get(id);

        collectedNews.push({
          id: id,
          title: item.title,
          summary: summary,
          sourceName: source.name,
          sourceUrl: item.url,
          sourceHome: source.url,
          category: category,
          publishedAt: previousItem?.publishedAt || nowStr,
          collectedAt: nowStr,
          imageUrl: item.imageUrl || previousItem?.imageUrl || "",
          tags: ["Niger", region, category],
          region: region,
          importance: importance,
          isSample: false
        });
      });

      source.lastChecked = nowStr;
      successfulSourcesCount++;

    } catch (err) {
      console.error(`[Source ${index}/${sources.length}] [Erreur] Échec pour ${source.name} : ${err.message}`);
    }
  }

  // 3. Exécution concurrente par lots (Pool de Concurrence)
  // Utiliser une taille de lot (batch size) de 5 requêtes simultanées
  const CONCURRENCY_LIMIT = 5;
  for (let i = 0; i < sources.length; i += CONCURRENCY_LIMIT) {
    const chunk = sources.slice(i, i + CONCURRENCY_LIMIT);
    console.log(`\n--- Lancement du lot de sources [${i + 1} à ${Math.min(i + CONCURRENCY_LIMIT, sources.length)}] ---`);
    await Promise.all(chunk.map((src, indexInChunk) => scrapeSource(src, i + indexInChunk + 1)));
  }

  // 4. Fusionner les nouvelles collectées avec les anciennes
  // Prioriser les nouveaux éléments
  let mergedNews = [...collectedNews, ...oldNews];

  // Dédupliquer par ID global
  const finalNews = [];
  const seenIds = new Set();
  
  for (const item of mergedNews) {
    if (!seenIds.has(item.id)) {
      seenIds.add(item.id);
      finalNews.push(item);
    }
  }

  // Trier par date collectée / publiée décroissante
  finalNews.sort((a, b) => new Date(b.publishedAt) - new Date(a.publishedAt));

  // Limiter à 1000 éléments au total
  const truncatedNews = finalNews.slice(0, MAX_TOTAL_ITEMS);

  // 5. Sauvegarder les fichiers mis à jour
  try {
    // Si nous n'avons rien collecté de nouveau (ex : tout a échoué par absence de réseau dans l'environnement),
    // nous écrivons quand même les sources avec leurs états mis à jour, et nous laissons les news intactes.
    fs.writeFileSync(PATH_NEWS, JSON.stringify(truncatedNews, null, 2), 'utf8');
    fs.writeFileSync(PATH_SOURCES, JSON.stringify(sources, null, 2), 'utf8');
    
    console.log('\n=== RAPPORT DE COLLECTE ===');
    console.log(`- Sources vérifiées avec succès : ${successfulSourcesCount}/${sources.length}`);
    console.log(`- Nouveaux articles collectés lors de cette session : ${collectedNews.length}`);
    console.log(`- Nombre total d'articles stockés dans data/news.json : ${truncatedNews.length}`);
    console.log('=============================================');
  } catch (writeErr) {
    console.error('Erreur lors de l\'écriture des fichiers mis à jour:', writeErr);
  }
}

// Lancer la collecte
if (process.argv.includes('--daemon')) {
  const intervalSeconds = 15;
  console.log(`[Mode Démon] Démarrage de la veille en continu (toutes les ${intervalSeconds} secondes)...`);
  
  // Exécuter immédiatement
  collect();
  
  // Planifier le cycle
  setInterval(() => {
    collect();
  }, intervalSeconds * 1000);
} else {
  // Mode exécution unique (standard ou GitHub Actions)
  collect();
}

File: scripts/collect_news.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Niger Info Veille - Collecteur d'Actualités Ultra-Puissant en Python
Caractéristiques : Multi-threadé (concurrence de 5), tolérant aux pannes,
extraction d'images, classification intelligente et support du mode démon.
"""

import os
import sys
import json
import time
import hashlib
import urllib.parse
from datetime import datetime, timedelta, timezone
from concurrent.futures import ThreadPoolExecutor, as_completed
import warnings

# Tenter d'importer les dépendances externes requises
try:
    import requests
    from bs4 import BeautifulSoup
except ImportError:
    print("Erreur : Les bibliothèques 'requests' et 'beautifulsoup4' sont requises.")
    print("Veuillez installer les dépendances en lançant : pip install requests beautifulsoup4")
    sys.exit(1)

# Désactiver les avertissements SSL pour les sites avec certificats expirés/invalides
from requests.packages.urllib3.exceptions import InsecureRequestWarning
requests.packages.urllib3.disable_warnings(InsecureRequestWarning)

# Configurations
USER_AGENT = "NigerInfoVeilleBot/1.0 (+contact@example.com)"
MAX_ITEMS_PER_SOURCE = 10
MAX_TOTAL_ITEMS = 1000
MIN_TITLE_LENGTH = 30
TIMEOUT_SECONDS = 10
CONCURRENCY_LIMIT = 5
IMAGE_ENRICH_LIMIT_PER_SOURCE = 6

# Répertoire de travail
BASE_DIR = os.path.dirname(os.path.abspath(__file__))
PATH_SOURCES = os.path.join(BASE_DIR, '../data/sources.json')
PATH_NEWS = os.path.join(BASE_DIR, '../data/news.json')

def fetch_html(url):
    """Télécharge le HTML d'une page avec gestion des redirections et des certificats invalides."""
    headers = {
        'User-Agent': USER_AGENT,
        'Accept': 'text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,*/*;q=0.8',
        'Accept-Language': 'fr,fr-FR;q=0.8,en-US;q=0.5,en;q=0.3'
    }
    
    # Premier essai avec vérification SSL stricte
    try:
        response = requests.get(url, headers=headers, timeout=TIMEOUT_SECONDS, allow_redirects=True)
        response.raise_for_status()
        return response.text
    except requests.exceptions.SSLError:
        # Deuxième essai sans vérification SSL (très utile pour les sites administratifs du Niger)
        response = requests.get(url, headers=headers, timeout=TIMEOUT_SECONDS, allow_redirects=True, verify=False)
        response.raise_for_status()
        return response.text

def normalize_image_url(raw_url, base_url):
    """Normalise une URL d'image et écarte logos, icônes et pixels de suivi."""
    if not raw_url or not isinstance(raw_url, str):
        return ""

    candidate = raw_url.split(',')[-1].strip().split()[0]
    if not candidate or candidate.startswith(('data:', 'blob:')):
        return ""

    absolute_url = urllib.parse.urljoin(base_url, candidate)
    lower_url = absolute_url.lower()
    rejected_tokens = [
        'logo', 'icon', 'favicon', 'avatar', 'sprite', 'pixel', 'tracking',
        'facebook', 'twitter', 'instagram', 'youtube', 'whatsapp'
    ]
    if any(token in lower_url for token in rejected_tokens) or lower_url.endswith('.svg'):
        return ""
    return absolute_url

def image_from_tag(img_tag, base_url):
    """Extrait la meilleure URL disponible d'une balise image classique ou lazy-loadée."""
    if not img_tag:
        return ""
    for attribute in [
        'data-src', 'data-original', 'data-lazy-src', 'data-image',
        'data-srcset', 'srcset', 'src'
    ]:
        image_url = normalize_image_url(img_tag.get(attribute), base_url)
        if image_url:
            return image_url
    return ""

def extract_page_image(html, page_url):
    """Extrait l'image éditoriale depuis Open Graph, Twitter Cards ou le contenu."""
    soup = BeautifulSoup(html, 'html.parser')
    meta_candidates = [
        ('meta', {'property': 'og:image:secure_url'}, 'content'),
        ('meta', {'property': 'og:image'}, 'content'),
        ('meta', {'name': 'twitter:image'}, 'content'),
        ('meta', {'name': 'twitter:image:src'}, 'content'),
        ('link', {'rel': 'image_src'}, 'href')
    ]

    for tag_name, attributes, value_attribute in meta_candidates:
        element = soup.find(tag_name, attrs=attributes)
        if element:
            image_url = normalize_image_url(element.get(value_attribute), page_url)
            if image_url:
                return image_url

    content_image = soup.select_one('article img, main img, .article img, .post img')
    return image_from_tag(content_image, page_url)

def generate_news_id(source_name, title, url):
    """Génère un ID MD5 unique basé sur la source, le titre et l'URL."""
    hash_input = f"{source_name}-{title}-{url}".encode('utf-8')
    md5_hash = hashlib.md5(hash_input).hexdigest()
    return f"news-{md5_hash[:12]}"

def detect_region(title):
    """Détecte la région du Niger mentionnée dans le titre."""
    title_lower = title.lower()
    regions = ['Niamey', 'Maradi', 'Zinder', 'Tahoua', 'Dosso', 'Tillabéri', 'Agadez', 'Diffa']
    for region in regions:
        if region.lower() in title_lower:
            return region
    return 'Niger'

def estimate_category(title, default_category):
    """Catégorise l'article selon des mots-clés présents dans le titre."""
    t = title.lower()
    if any(k in t for k in ["appel d'offres", "marché public", "avis d'attribution", "soumission"]):
        return "Marchés publics"
    if any(k in t for k in ["santé", "hôpital", "vaccin", "épidémie", "maladie", "polio"]):
        return "Santé"
    if any(k in t for k in ["pluie", "météo", "climat", "inondation", "sécheresse", "tempête"]):
        return "Météo"
    if any(k in t for k in ["agriculture", "semence", "élevage", "bétail", "mil ", "céréale"]):
        return "Agriculture"
    if any(k in t for k in ["budget", "finance", "économie", "banque", "bceao", "fmi", "inflation"]):
        return "Économie"
    if any(k in t for k in ["école", "université", "éducation", "classe", "scolaire"]):
        return "Éducation"
    if any(k in t for k in ["humanitaire", "déplacés", "nutrition", "réfugiés", "aide alimentaire", "ocha"]):
        return "Humanitaire"
    if any(k in t for k in ["sécurité", "militaire", "attaque", "défense", "soldat", "terrorisme"]):
        return "Sécurité"
    if any(k in t for k in ["communiqué officiel", "conseil des ministres", "décret", "gouvernement"]):
        return "Gouvernement"
    if any(k in t for k in ["diaspora", "ambassade", "consulat"]):
        return "Diaspora"
    return default_category

def estimate_importance(title, source_name, category):
    """Estime le degré d'importance de l'article (1 à 5)."""
    t = title.lower()
    if any(k in t for k in ["urgent", "alerte rouge", "inondation majeure", "attaque terroriste", "coup d'état"]):
        return 5
    if any(k in t for k in ["communiqué", "décret", "alerte", "sécurité", "conseil des ministres"]):
        return 4
    if category == "Gouvernement" or category == "Marchés publics" or "présidence" in source_name.lower() or "gouvernement" in source_name.lower():
        return 3
    return 2

def scrape_source(source, index, total_sources, old_news_by_id):
    """Analyse une source unique pour extraire les actualités et les images."""
    if source.get('status') != 'active':
        print(f"[Source {index}/{total_sources}] Ignorée : {source['name']} (inactive)")
        return []

    print(f"[Source {index}/{total_sources}] Début de l'analyse : {source['name']} ({source['url']})")
    
    collected_items = []
    try:
        html = fetch_html(source['url'])
        soup = BeautifulSoup(html, 'html.parser')
        source_items = []
        
        # Analyser toutes les balises <a>
        for link in soup.find_all('a'):
            href = link.get('href')
            text = ' '.join(link.get_text().split()).strip()
            
            if not href or not text:
                continue
            if len(text) < MIN_TITLE_LENGTH:
                continue
                
            text_lower = text.lower()
            blacklist = [
                'accueil', 'contact', 'menu', 'facebook', 'twitter', 'linkedin', 'instagram', 'youtube',
                'lire plus', 'lire l\'article', 'read more', 'connexion', 's\'abonner', 'newsletter',
                'partager', 'télécharger', 'mentions légales', 'politique de confidentialité', 'qui sommes-nous'
            ]
            if any(text_lower.startswith(b) or text_lower == b for b in blacklist):
                continue
                
            if href.startswith('mailto:') or href.startswith('tel:') or href.startswith('javascript:') or href.startswith('#'):
                continue
                
            # Résoudre l'URL absolue
            absolute_url = urllib.parse.urljoin(source['url'], href)
            
            # Tenter d'extraire une image associée
            image_url = ""
            img_tag = link.find('img')
            image_url = image_from_tag(img_tag, source['url'])

            if not image_url:
                card = link.find_parent(['article', 'li'])
                if not card:
                    card = link.find_parent(class_=['card', 'post', 'news-item', 'item', 'entry'])
                if card:
                    image_url = image_from_tag(card.find('img'), source['url'])
                    
            source_items.append({
                'title': text,
                'url': absolute_url,
                'image_url': image_url
            })
            
        # Dédupliquer les liens sur la même page
        unique_items = []
        seen_urls = set()
        for item in source_items:
            if item['url'] not in seen_urls:
                seen_urls.add(item['url'])
                unique_items.append(item)
                
        # Limiter par source
        selected_items = unique_items[:MAX_ITEMS_PER_SOURCE]

        for item in [entry for entry in selected_items if not entry['image_url']][:IMAGE_ENRICH_LIMIT_PER_SOURCE]:
            try:
                article_html = fetch_html(item['url'])
                item['image_url'] = extract_page_image(article_html, item['url'])
            except Exception:
                # Une image indisponible ne doit pas bloquer l'article.
                pass

        print(f"[Source {index}/{total_sources}] -> {len(selected_items)} liens trouvés pour {source['name']}")
        
        now_str = datetime.now(timezone(timedelta(hours=1))).isoformat()
        for item in selected_items:
            news_id = generate_news_id(source['name'], item['title'], item['url'])
            region = detect_region(item['title'])
            category = estimate_category(item['title'], source['category'])
            importance = estimate_importance(item['title'], source['name'], category)
            summary = f"Cette information a été repérée depuis {source['name']}. Elle concerne : {item['title']}. Consultez la source originale pour lire le contenu complet."
            previous_item = old_news_by_id.get(news_id, {})
            
            collected_items.append({
                "id": news_id,
                "title": item['title'],
                "summary": summary,
                "sourceName": source['name'],
                "sourceUrl": item['url'],
                "sourceHome": source['url'],
                "category": category,
                "publishedAt": previous_item.get('publishedAt', now_str),
                "collectedAt": now_str,
                "imageUrl": item['image_url'] or previous_item.get('imageUrl', ''),
                "tags": ["Niger", region, category],
                "region": region,
                "importance": importance,
                "isSample": False
            })
            
        # Marquer la source comme vérifiée
        source['lastChecked'] = now_str
        
    except Exception as e:
        print(f"[Source {index}/{total_sources}] [Erreur] Échec pour {source['name']} : {str(e)}")
        
    return collected_items

def generate_rss_feed(news_items):
    """Génère un flux RSS 2.0 valide contenant les 50 actualités les plus récentes."""
    rss_path = os.path.join(BASE_DIR, '../data/rss.xml')
    items_xml = []
    
    for item in news_items[:50]:
        pub_date = item['publishedAt']
        try:
            # Essayer de formater en RFC 822 pour la compatibilité RSS
            dt = datetime.fromisoformat(pub_date.replace('Z', '+00:00'))
            pub_date_formatted = dt.strftime('%a, %d %b %Y %H:%M:%S %z')
        except Exception:
            pub_date_formatted = pub_date
            
        items_xml.append(f"""    <item>
      <title><![CDATA[{item['title']}]]></title>
      <link>{item['sourceUrl']}</link>
      <guid isPermaLink="false">{item['id']}</guid>
      <pubDate>{pub_date_formatted}</pubDate>
      <description><![CDATA[{item['summary']}]]></description>
      <source url="{item['sourceHome']}">{item['sourceName']}</source>
      <category>{item['category']}</category>
    </item>""")
        
    rss_content = f"""<?xml version="1.0" encoding="UTF-8" ?>
<rss version="2.0" xmlns:atom="http://www.w3.org/2005/Atom">
<channel>
  <title>Niger Info Veille - Flux RSS</title>
  <link>https://issou2025.github.io/nigerinfo/</link>
  <description>Flux d'actualités et de communiqués officiels sur le Niger.</description>
  <language>fr</language>
  <lastBuildDate>{datetime.now().strftime('%a, %d %b %Y %H:%M:%S +0100')}</lastBuildDate>
  <atom:link href="https://issou2025.github.io/nigerinfo/data/rss.xml" rel="self" type="application/rss+xml" />
{chr(10).join(items_xml)}
</channel>
</rss>"""
    
    with open(rss_path, 'w', encoding='utf-8') as f:
        f.write(rss_content)
    print("Flux RSS statique généré avec succès dans data/rss.xml")

def collect():
    """Charge les sources, exécute le scan multi-threadé, fusionne et écrit les résultats."""
    print("=== DÉBUT DE LA COLLECTE NIGER INFO VEILLE (PYTHON ENGINE) ===")
    
    if not os.path.exists(PATH_SOURCES):
        print(f"Fichier sources.json introuvable à : {PATH_SOURCES}")
        sys.exit(1)
        
    with open(PATH_SOURCES, 'r', encoding='utf-8') as f:
        sources = json.load(f)
    print(f"{len(sources)} sources chargées.")
    
    old_news = []
    if os.path.exists(PATH_NEWS):
        try:
            with open(PATH_NEWS, 'r', encoding='utf-8') as f:
                old_news = json.load(f)
        except Exception:
            old_news = []
    print(f"{len(old_news)} articles actuellement en cache local.")
    old_news_by_id = {item['id']: item for item in old_news}
    
    collected_news = []
    total_sources = len(sources)
    
    # Exécuter les requêtes en parallèle (ThreadPoolExecutor) pour un scan très rapide
    with ThreadPoolExecutor(max_workers=CONCURRENCY_LIMIT) as executor:
        futures = {
            executor.submit(scrape_source, src, idx + 1, total_sources, old_news_by_id): src
            for idx, src in enumerate(sources)
        }
        for future in as_completed(futures):
            results = future.result()
            if results:
                collected_news.extend(results)
                
    # Fusionner et dédupliquer
    merged_news = collected_news + old_news
    final_news = []
    seen_ids = set()
    for item in merged_news:
        if item['id'] not in seen_ids:
            seen_ids.add(item['id'])
            final_news.append(item)
            
    # Trier par date décroissante
    final_news.sort(key=lambda x: x['publishedAt'], reverse=True)
    
    # Tranchage
    truncated_news = final_news[:MAX_TOTAL_ITEMS]
    
    # Compter les sources vérifiées avec succès
    successful_sources = sum(1 for src in sources if src.get('lastChecked'))
    
    # Enregistrer
    try:
        with open(PATH_NEWS, 'w', encoding='utf-8') as f:
            json.dump(truncated_news, f, ensure_ascii=False, indent=2)
        with open(PATH_SOURCES, 'w', encoding='utf-8') as f:
            json.dump(sources, f, ensure_ascii=False, indent=2)
            
        # Générer le flux RSS statique
        generate_rss_feed(truncated_news)
            
        print('\n=== RAPPORT DE COLLECTE PYTHON ===')
        print(f"- Sources scannées avec succès : {successful_sources}/{total_sources}")
        print(f"- Nouveaux articles récoltés lors de cette passe : {len(collected_news)}")
        print(f"- Total articles dans le fichier news.json : {len(truncated_news)}")
        print('=============================================')
    except Exception as e:
        print(f"Erreur d'écriture des fichiers : {str(e)}")

if __name__ == '__main__':
    # Gestion du mode démon
    if '--daemon' in sys.argv:
        interval_seconds = 15
        print(f"[Mode Démon Python] Démarrage de la veille en continu (toutes les {interval_seconds}s)...")
        while True:
            collect()
            time.sleep(interval_seconds)
    else:
        collect()
